From 226292434e03d182c25beb453045139b6a34c7b5 Mon Sep 17 00:00:00 2001
From: Michal Kubecek <mkubecek@suse.cz>
Date: Sun, 10 Mar 2019 20:41:44 +0100
Subject: [PATCH 27/44] ethtool: provide coalescing information in PARAMS_GET
 request

Introduce PARAMS_GET netlink request to get coalescing information as
provided by ETHTOOL_GCOALESCE ioctl command if ETHTOOL_IM_PARAMS_COALESCE
flag is set in the request.

Signed-off-by: Michal Kubecek <mkubecek@suse.cz>
---
 Documentation/networking/ethtool-netlink.txt |  51 +++++-
 include/uapi/linux/ethtool_netlink.h         |  50 ++++++
 net/ethtool/Makefile                         |   2 +-
 net/ethtool/netlink.c                        |   8 +
 net/ethtool/netlink.h                        |   1 +
 net/ethtool/params.c                         | 163 +++++++++++++++++++
 6 files changed, 273 insertions(+), 2 deletions(-)
 create mode 100644 net/ethtool/params.c

diff --git a/Documentation/networking/ethtool-netlink.txt b/Documentation/networking/ethtool-netlink.txt
index ea6bb6daa71a..9c5eb8f7dd46 100644
--- a/Documentation/networking/ethtool-netlink.txt
+++ b/Documentation/networking/ethtool-netlink.txt
@@ -154,6 +154,7 @@ Userspace to kernel:
     ETHTOOL_MSG_SETTINGS_GET		get device settings
     ETHTOOL_MSG_SETTINGS_SET		set device settings
     ETHTOOL_MSG_INFO_GET		get device information
+    ETHTOOL_MSG_PARAMS_GET		get device parameters
 
 Kernel to userspace:
 
@@ -162,6 +163,7 @@ Kernel to userspace:
     ETHTOOL_MSG_SETTINGS_NTF		device settings notification
     ETHTOOL_MSG_SETTINGS_SET_REPLY	additional reply for SETTINGS_SET
     ETHTOOL_MSG_INFO_GET_REPLY		device information
+    ETHTOOL_MSG_PARAMS_GET_REPLY	device parameters
 
 "GET" requests are sent by userspace applications to retrieve device
 information. They usually do not contain any message specific attributes.
@@ -429,6 +431,53 @@ value; if no PHC is associated, the attribute is not present.
 INFO_GET requests allow dumps.
 
 
+PARAMS_GET
+----------
+
+PARAMS_GET request retrieves information provided by ioctl comand
+ETHTOOL_GCOALESCE (coalescing parameters).
+
+Request contents:
+
+    ETHTOOL_A_PARAMS_HEADER		(nested)	request header
+
+Info mask bits:
+
+    ETHTOOL_IM_PARAMS_COALESCE		coalescing parameters
+
+Response contents: On top level, there is one attribute for each of the
+information categories, the information is nested in it.
+
+    ETHTOOL_A_PARAMS_HEADER		(nested)	reply header
+    ETHTOOL_A_PARAMS_COALESCE		(nested)	coalescing parameters
+        ETHTOOL_A_COALESCE_RX_USECS	    (u32)
+        ETHTOOL_A_COALESCE_RX_MAXFRM	    (u32)
+        ETHTOOL_A_COALESCE_RX_USECS_IRQ	    (u32)
+        ETHTOOL_A_COALESCE_RX_MAXFRM_IRQ    (u32)
+        ETHTOOL_A_COALESCE_RX_USECS_LOW	    (u32)
+        ETHTOOL_A_COALESCE_RX_MAXFRM_LOW    (u32)
+        ETHTOOL_A_COALESCE_RX_USECS_HIGH    (u32)
+        ETHTOOL_A_COALESCE_RX_MAXFRM_HIGH   (u32)
+        ETHTOOL_A_COALESCE_TX_USECS	    (u32)
+        ETHTOOL_A_COALESCE_TX_MAXFRM	    (u32)
+        ETHTOOL_A_COALESCE_TX_USECS_IRQ	    (u32)
+        ETHTOOL_A_COALESCE_TX_MAXFRM_IRQ    (u32)
+        ETHTOOL_A_COALESCE_TX_USECS_LOW	    (u32)
+        ETHTOOL_A_COALESCE_TX_MAXFRM_LOW    (u32)
+        ETHTOOL_A_COALESCE_TX_USECS_HIGH    (u32)
+        ETHTOOL_A_COALESCE_TX_MAXFRM_HIGH   (u32)
+        ETHTOOL_A_COALESCE_PKT_RATE_LOW	    (u32)
+        ETHTOOL_A_COALESCE_PKT_RATE_HIGH    (u32)
+        ETHTOOL_A_COALESCE_RX_USE_ADAPTIVE  (bool)
+        ETHTOOL_A_COALESCE_TX_USE_ADAPTIVE  (bool)
+        ETHTOOL_A_COALESCE_RATE_SAMPLE_INTERVAL	(u32)
+        ETHTOOL_A_COALESCE_STATS_BLOCK_USECS	(u32)
+
+PARAMS_GET requests allow dumps and messages in the same format as response
+to them are broadcasted as notifications on change of these settings using
+netlink or ioctl ethtool interface.
+
+
 Request translation
 -------------------
 
@@ -450,7 +499,7 @@ ETHTOOL_NWAY_RST		n/a
 ETHTOOL_GLINK			ETHTOOL_MSG_SETTINGS_GET
 ETHTOOL_GEEPROM			n/a
 ETHTOOL_SEEPROM			n/a
-ETHTOOL_GCOALESCE		n/a
+ETHTOOL_GCOALESCE		ETHTOOL_MSG_PARAMS_GET
 ETHTOOL_SCOALESCE		n/a
 ETHTOOL_GRINGPARAM		n/a
 ETHTOOL_SRINGPARAM		n/a
diff --git a/include/uapi/linux/ethtool_netlink.h b/include/uapi/linux/ethtool_netlink.h
index 4d952abbfab3..b04a77849ff5 100644
--- a/include/uapi/linux/ethtool_netlink.h
+++ b/include/uapi/linux/ethtool_netlink.h
@@ -18,6 +18,7 @@ enum {
 	ETHTOOL_MSG_SETTINGS_GET,
 	ETHTOOL_MSG_SETTINGS_SET,
 	ETHTOOL_MSG_INFO_GET,
+	ETHTOOL_MSG_PARAMS_GET,
 
 	/* add new constants above here */
 	__ETHTOOL_MSG_USER_CNT,
@@ -32,6 +33,7 @@ enum {
 	ETHTOOL_MSG_SETTINGS_NTF,
 	ETHTOOL_MSG_SETTINGS_SET_REPLY,
 	ETHTOOL_MSG_INFO_GET_REPLY,
+	ETHTOOL_MSG_PARAMS_GET_REPLY,
 
 	/* add new constants above here */
 	__ETHTOOL_MSG_KERNEL_CNT,
@@ -301,6 +303,54 @@ enum {
 	ETHTOOL_A_TSINFO_MAX = (__ETHTOOL_A_TSINFO_CNT - 1)
 };
 
+/* GET_PARAMS / SET_PARAMS */
+
+enum {
+	ETHTOOL_A_PARAMS_UNSPEC,
+	ETHTOOL_A_PARAMS_HEADER,		/* nest - _A_HEADER_* */
+	ETHTOOL_A_PARAMS_COALESCE,		/* nest - _A_COALESCE_* */
+
+	/* add new constants above here */
+	__ETHTOOL_A_PARAMS_CNT,
+	ETHTOOL_A_PARAMS_MAX = (__ETHTOOL_A_PARAMS_CNT - 1)
+};
+
+#define ETHTOOL_IM_PARAMS_COALESCE		(1U << 0)
+
+#define ETHTOOL_IM_PARAMS_ALL (ETHTOOL_IM_PARAMS_COALESCE)
+
+#define ETHTOOL_RFLAG_PARAMS_ALL 0
+
+enum {
+	ETHTOOL_A_COALESCE_UNSPEC,
+	ETHTOOL_A_COALESCE_RX_USECS,		/* u32 */
+	ETHTOOL_A_COALESCE_RX_MAXFRM,		/* u32 */
+	ETHTOOL_A_COALESCE_RX_USECS_IRQ,	/* u32 */
+	ETHTOOL_A_COALESCE_RX_MAXFRM_IRQ,	/* u32 */
+	ETHTOOL_A_COALESCE_RX_USECS_LOW,	/* u32 */
+	ETHTOOL_A_COALESCE_RX_MAXFRM_LOW,	/* u32 */
+	ETHTOOL_A_COALESCE_RX_USECS_HIGH,	/* u32 */
+	ETHTOOL_A_COALESCE_RX_MAXFRM_HIGH,	/* u32 */
+	ETHTOOL_A_COALESCE_TX_USECS,		/* u32 */
+	ETHTOOL_A_COALESCE_TX_MAXFRM,		/* u32 */
+	ETHTOOL_A_COALESCE_TX_USECS_IRQ,	/* u32 */
+	ETHTOOL_A_COALESCE_TX_MAXFRM_IRQ,	/* u32 */
+	ETHTOOL_A_COALESCE_TX_USECS_LOW,	/* u32 */
+	ETHTOOL_A_COALESCE_TX_MAXFRM_LOW,	/* u32 */
+	ETHTOOL_A_COALESCE_TX_USECS_HIGH,	/* u32 */
+	ETHTOOL_A_COALESCE_TX_MAXFRM_HIGH,	/* u32 */
+	ETHTOOL_A_COALESCE_PKT_RATE_LOW,	/* u32 */
+	ETHTOOL_A_COALESCE_PKT_RATE_HIGH,	/* u32 */
+	ETHTOOL_A_COALESCE_RX_USE_ADAPTIVE,	/* u8 */
+	ETHTOOL_A_COALESCE_TX_USE_ADAPTIVE,	/* u8 */
+	ETHTOOL_A_COALESCE_RATE_SAMPLE_INTERVAL, /* u32 */
+	ETHTOOL_A_COALESCE_STATS_BLOCK_USECS,	/* u32 */
+
+	/* add new constants above here */
+	__ETHTOOL_A_COALESCE_CNT,
+	ETHTOOL_A_COALESCE_MAX = (__ETHTOOL_A_COALESCE_CNT - 1)
+};
+
 /* generic netlink info */
 #define ETHTOOL_GENL_NAME "ethtool"
 #define ETHTOOL_GENL_VERSION 1
diff --git a/net/ethtool/Makefile b/net/ethtool/Makefile
index 07e824c5eafa..c886f3d2c184 100644
--- a/net/ethtool/Makefile
+++ b/net/ethtool/Makefile
@@ -4,4 +4,4 @@ obj-y				+= ioctl.o common.o
 
 obj-$(CONFIG_ETHTOOL_NETLINK)	+= ethtool_nl.o
 
-ethtool_nl-y	:= netlink.o bitset.o strset.o settings.o info.o
+ethtool_nl-y	:= netlink.o bitset.o strset.o settings.o info.o params.o
diff --git a/net/ethtool/netlink.c b/net/ethtool/netlink.c
index 090cebe8e3ab..b3261087fd1b 100644
--- a/net/ethtool/netlink.c
+++ b/net/ethtool/netlink.c
@@ -297,6 +297,7 @@ static const struct get_request_ops *get_requests[__ETHTOOL_MSG_USER_CNT] = {
 	[ETHTOOL_MSG_STRSET_GET]	= &strset_request_ops,
 	[ETHTOOL_MSG_SETTINGS_GET]	= &settings_request_ops,
 	[ETHTOOL_MSG_INFO_GET]		= &info_request_ops,
+	[ETHTOOL_MSG_PARAMS_GET]	= &params_request_ops,
 };
 
 static struct ethnl_dump_ctx *ethnl_dump_context(struct netlink_callback *cb)
@@ -759,6 +760,13 @@ static const struct genl_ops ethtool_genl_ops[] = {
 		.dumpit	= ethnl_get_dumpit,
 		.done	= ethnl_get_done,
 	},
+	{
+		.cmd	= ETHTOOL_MSG_PARAMS_GET,
+		.doit	= ethnl_get_doit,
+		.start	= ethnl_get_start,
+		.dumpit	= ethnl_get_dumpit,
+		.done	= ethnl_get_done,
+	},
 };
 
 static const struct genl_multicast_group ethtool_nl_mcgrps[] = {
diff --git a/net/ethtool/netlink.h b/net/ethtool/netlink.h
index b01c9ca6637f..647d47715bde 100644
--- a/net/ethtool/netlink.h
+++ b/net/ethtool/netlink.h
@@ -360,6 +360,7 @@ struct get_request_ops {
 extern const struct get_request_ops strset_request_ops;
 extern const struct get_request_ops settings_request_ops;
 extern const struct get_request_ops info_request_ops;
+extern const struct get_request_ops params_request_ops;
 
 int ethnl_set_settings(struct sk_buff *skb, struct genl_info *info);
 
diff --git a/net/ethtool/params.c b/net/ethtool/params.c
new file mode 100644
index 000000000000..4e521a85b434
--- /dev/null
+++ b/net/ethtool/params.c
@@ -0,0 +1,163 @@
+/* SPDX-License-Identifier: GPL-2.0 WITH Linux-syscall-note */
+
+#include "netlink.h"
+
+static const struct nla_policy params_get_policy[ETHTOOL_A_PARAMS_MAX + 1] = {
+	[ETHTOOL_A_PARAMS_UNSPEC]		= { .type = NLA_REJECT },
+	[ETHTOOL_A_PARAMS_HEADER]		= { .type = NLA_NESTED },
+	[ETHTOOL_A_PARAMS_COALESCE]		= { .type = NLA_REJECT },
+};
+
+struct params_data {
+	struct ethnl_req_info		reqinfo_base;
+
+	/* everything below here will be reset for each device in dumps */
+	struct ethnl_reply_data		repdata_base;
+	struct ethtool_coalesce		coalesce;
+};
+
+static int ethnl_get_coalesce(struct net_device *dev,
+			      struct ethtool_coalesce *data)
+{
+	if (!dev->ethtool_ops->get_coalesce)
+		return -EOPNOTSUPP;
+	return dev->ethtool_ops->get_coalesce(dev, data);
+}
+
+static int params_prepare(struct ethnl_req_info *req_info,
+			  struct genl_info *info)
+{
+	struct params_data *data =
+		container_of(req_info, struct params_data, reqinfo_base);
+	struct net_device *dev = data->repdata_base.dev;
+	u32 req_mask = req_info->req_mask;
+	int ret;
+
+	ret = ethnl_before_ops(dev);
+	if (ret < 0)
+		return ret;
+	if (req_mask & ETHTOOL_IM_PARAMS_COALESCE) {
+		ret = ethnl_get_coalesce(dev, &data->coalesce);
+		if (ret < 0)
+			req_mask &= ~ETHTOOL_IM_PARAMS_COALESCE;
+	}
+	ethnl_after_ops(dev);
+
+	data->repdata_base.info_mask = req_mask;
+	if ((req_info->req_mask & ~req_mask) && info)
+		GENL_SET_ERR_MSG(info, "not all requested data could be retrieved");
+	return 0;
+}
+
+static int params_coalesce_size(void)
+{
+	return nla_total_size(20 * nla_total_size(sizeof(u32)) +
+			      2 * nla_total_size(sizeof(u8)));
+}
+
+static int params_size(const struct ethnl_req_info *req_info)
+{
+	struct params_data *data =
+		container_of(req_info, struct params_data, reqinfo_base);
+	u32 info_mask = data->repdata_base.info_mask;
+	int len = 0;
+
+	len += ethnl_reply_header_size();
+	if (info_mask & ETHTOOL_IM_PARAMS_COALESCE)
+		len += params_coalesce_size();
+
+	return len;
+}
+
+static int params_fill_coalesce(struct sk_buff *skb,
+				struct ethtool_coalesce *data)
+{
+	struct nlattr *nest;
+
+	nest = nla_nest_start(skb, ETHTOOL_A_PARAMS_COALESCE);
+	if (!nest)
+		return -EMSGSIZE;
+	if (nla_put_u32(skb, ETHTOOL_A_COALESCE_RX_USECS,
+			data->rx_coalesce_usecs) ||
+	    nla_put_u32(skb, ETHTOOL_A_COALESCE_RX_MAXFRM,
+			data->rx_max_coalesced_frames) ||
+	    nla_put_u32(skb, ETHTOOL_A_COALESCE_RX_USECS_IRQ,
+			data->rx_coalesce_usecs_irq) ||
+	    nla_put_u32(skb, ETHTOOL_A_COALESCE_RX_MAXFRM_IRQ,
+			data->rx_max_coalesced_frames_irq) ||
+	    nla_put_u32(skb, ETHTOOL_A_COALESCE_RX_USECS_LOW,
+			data->rx_coalesce_usecs_low) ||
+	    nla_put_u32(skb, ETHTOOL_A_COALESCE_RX_MAXFRM_LOW,
+			data->rx_max_coalesced_frames_low) ||
+	    nla_put_u32(skb, ETHTOOL_A_COALESCE_RX_USECS_HIGH,
+			data->rx_coalesce_usecs_high) ||
+	    nla_put_u32(skb, ETHTOOL_A_COALESCE_RX_MAXFRM_HIGH,
+			data->rx_max_coalesced_frames_high) ||
+	    nla_put_u32(skb, ETHTOOL_A_COALESCE_TX_USECS,
+			data->tx_coalesce_usecs) ||
+	    nla_put_u32(skb, ETHTOOL_A_COALESCE_TX_MAXFRM,
+			data->tx_max_coalesced_frames) ||
+	    nla_put_u32(skb, ETHTOOL_A_COALESCE_TX_USECS_IRQ,
+			data->tx_coalesce_usecs_irq) ||
+	    nla_put_u32(skb, ETHTOOL_A_COALESCE_TX_MAXFRM_IRQ,
+			data->tx_max_coalesced_frames_irq) ||
+	    nla_put_u32(skb, ETHTOOL_A_COALESCE_TX_USECS_LOW,
+			data->tx_coalesce_usecs_low) ||
+	    nla_put_u32(skb, ETHTOOL_A_COALESCE_TX_MAXFRM_LOW,
+			data->tx_max_coalesced_frames_low) ||
+	    nla_put_u32(skb, ETHTOOL_A_COALESCE_TX_USECS_HIGH,
+			data->tx_coalesce_usecs_high) ||
+	    nla_put_u32(skb, ETHTOOL_A_COALESCE_TX_MAXFRM_HIGH,
+			data->tx_max_coalesced_frames_high) ||
+	    nla_put_u32(skb, ETHTOOL_A_COALESCE_PKT_RATE_LOW,
+			data->pkt_rate_low) ||
+	    nla_put_u32(skb, ETHTOOL_A_COALESCE_PKT_RATE_HIGH,
+			data->pkt_rate_high) ||
+	    nla_put_u8(skb, ETHTOOL_A_COALESCE_RX_USE_ADAPTIVE,
+		       !!data->use_adaptive_rx_coalesce) ||
+	    nla_put_u8(skb, ETHTOOL_A_COALESCE_TX_USE_ADAPTIVE,
+		       !!data->use_adaptive_tx_coalesce) ||
+	    nla_put_u32(skb, ETHTOOL_A_COALESCE_RATE_SAMPLE_INTERVAL,
+			data->rate_sample_interval) ||
+	    nla_put_u32(skb, ETHTOOL_A_COALESCE_STATS_BLOCK_USECS,
+			data->stats_block_coalesce_usecs)) {
+		nla_nest_cancel(skb, nest);
+		return -EMSGSIZE;
+	}
+
+	nla_nest_end(skb, nest);
+	return 0;
+}
+
+static int params_fill(struct sk_buff *skb,
+		       const struct ethnl_req_info *req_info)
+{
+	struct params_data *data =
+		container_of(req_info, struct params_data, reqinfo_base);
+	u32 info_mask = data->repdata_base.info_mask;
+	int ret;
+
+	if (info_mask & ETHTOOL_IM_PARAMS_COALESCE) {
+		ret = params_fill_coalesce(skb, &data->coalesce);
+		if (ret < 0)
+			return ret;
+	}
+
+	return 0;
+}
+
+const struct get_request_ops params_request_ops = {
+	.request_cmd		= ETHTOOL_MSG_PARAMS_GET,
+	.reply_cmd		= ETHTOOL_MSG_PARAMS_GET_REPLY,
+	.hdr_attr		= ETHTOOL_A_PARAMS_HEADER,
+	.max_attr		= ETHTOOL_A_PARAMS_MAX,
+	.data_size		= sizeof(struct params_data),
+	.repdata_offset		= offsetof(struct params_data, repdata_base),
+	.request_policy		= params_get_policy,
+	.default_infomask	= ETHTOOL_IM_PARAMS_ALL,
+	.all_reqflags		= ETHTOOL_RFLAG_PARAMS_ALL,
+
+	.prepare_data		= params_prepare,
+	.reply_size		= params_size,
+	.fill_reply		= params_fill,
+};
-- 
2.22.0

